base/broadcast.jl

```diff
@@ -137,11 +137,12 @@ BroadcastStyle(a::AbstractArrayStyle, ::Style{Tuple})    = a
 BroadcastStyle(::A, ::A) where A<:ArrayStyle             = A()
 BroadcastStyle(::ArrayStyle, ::ArrayStyle)               = Unknown()
 BroadcastStyle(::A, ::A) where A<:AbstractArrayStyle     = A()
-Base.@pure function BroadcastStyle(a::A, b::B) where {A<:AbstractArrayStyle{M},B<:AbstractArrayStyle{N}} where {M,N}
-    if Base.typename(A) === Base.typename(B)
-        return A(Val(max(M, N)))
+function BroadcastStyle(::A, ::B) where {M,N,A<:AbstractArrayStyle{M},B<:AbstractArrayStyle{N}}
+    if M!=N && M!=Any && N!=Any && A(Val(max(M,N)))==B(Val(max(M,N)))
```

Comment:
I think using `!==` is more compiler-friendly

```suggestion
    if M !== Any && N !== Any && M != N && A(Val(max(M,N))) == B(Val(max(M,N)))
```

(also Base code usually have spaces around the operators)

Alternatively, I guess defining

```julia
BroadcastStyle(::AbstractArrayStyle{Any}, ::AbstractArrayStyle{Any}) = Unknown()
BroadcastStyle(::AbstractArrayStyle, ::AbstractArrayStyle{Any}) = Unknown()
BroadcastStyle(::AbstractArrayStyle{Any}, ::AbstractArrayStyle) = Unknown()
```

would be the safest in terms of inferability. I think using `if` would be OK, though.

Reply:
Good point, thanks. I would keep tje `if` given that the code is clearer and if inference ever breaks here that test would catch it.